File: server/common/protocol.py
import logging

# Protocol constants
FIELD_SEPARATOR = 0x00    # Null byte separator between fields
BET_SEPARATOR = 0xFF      # Separator between different bets
PAYLOAD_LENGTH_BYTES = 4  # 4 bytes for the length of the message
MAX_MESSAGE_SIZE = 1024 * 1024 * 2 # 2MB
RESPONSE_HEADER_SIZE = 2 # 2 bytes for the success flag and separator

class BetData:
    """Represents a lottery bet in binary format"""
    
    def __init__(self, nombre="", apellido="", dni="", nacimiento="", numero=""):
        self.nombre = nombre
        self.apellido = apellido
        self.dni = dni
        self.nacimiento = nacimiento
        self.numero = numero


class Response:
    """Represents server response"""
    
    def __init__(self, success=False, message=""):
        self.success = success
        self.message = message


class Protocol:
    """Handles binary communication protocol"""
    
    def __init__(self):
        pass
    
    def _int_to_bytes(self, value):
        """Convert integer to 4 bytes in big-endian format"""
        return bytes([
            (value >> 24) & 0xFF,
            (value >> 16) & 0xFF,
            (value >> 8) & 0xFF,
            value & 0xFF
        ])
    
    def _bytes_to_int(self, data):
        """Convert 4 bytes to integer in big-endian format"""
        return (data[0] << 24) | (data[1] << 16) | (data[2] << 8) | data[3]
    
    def _recv_exact(self, sock, n):
        """Receive exactly n bytes or return None if connection closed."""
        buf = bytearray()
        while len(buf) < n:
            chunk = sock.recv(n - len(buf))
            if not chunk:   # peer closed before full data
                return None
            buf.extend(chunk)
        return bytes(buf)
    
    def deserialize_bet(self, data):
        """Convert binary data to BetData"""
        if len(data) < 1: # At least 1 byte for the first field
            raise ValueError("data too short")
        
        bet = BetData()
        offset = 0
        
        # Read nombre
        nombre_end = self._find_next_generic_separator(data, offset, FIELD_SEPARATOR)
        if nombre_end == -1:
            raise ValueError("invalid nombre field")
        bet.nombre = data[offset:nombre_end].decode('utf-8')
        offset = nombre_end + 1
        
        # Read apellido
        apellido_end = self._find_next_generic_separator(data, offset, FIELD_SEPARATOR)
        if apellido_end == -1:
            raise ValueError("invalid apellido field")
        bet.apellido = data[offset:apellido_end].decode('utf-8')
        offset = apellido_end + 1
        
        # Read dni
        dni_end = self._find_next_generic_separator(data, offset, FIELD_SEPARATOR)
        if dni_end == -1:
            raise ValueError("invalid dni field")
        bet.dni = data[offset:dni_end].decode('utf-8')
        offset = dni_end + 1
        
        # Read nacimiento
        nacimiento_end = self._find_next_generic_separator(data, offset, FIELD_SEPARATOR)
        if nacimiento_end == -1:
            raise ValueError("invalid nacimiento field")
        bet.nacimiento = data[offset:nacimiento_end].decode('utf-8')
        offset = nacimiento_end + 1
        
        # Read numero (last field, no separator)
        bet.numero = data[offset:].decode('utf-8')
        
        return bet
    
    def deserialize_batch(self, data):
        """Convert binary data to a list of BetData"""
        bets = []
        offset = 0
        
        while offset < len(data):
            # Find the end of the current bet (next bet separator)
            bet_end = self._find_next_generic_separator(data, offset, BET_SEPARATOR)
            if bet_end == -1:
                # No bet separator found, use remaining data
                bet_end = len(data)
            
            # Extract single bet data
            bet_data = data[offset:bet_end]
            if len(bet_data) > 0:
                try:
                    bet = self.deserialize_bet(bet_data)
                    bets.append(bet)
                except ValueError as e:
                    # If we can't parse a bet, we'll return what we have so far
                    # and let the caller handle the error
                    break
            
            # Move to next bet (skip the bet separator)
            offset = bet_end + 1 if bet_end < len(data) else bet_end
        
        return bets
    
    def _find_next_generic_separator(self, data, offset, separator):
        """Find the next generic separator starting from offset"""
        for i in range(offset, len(data)):
            if data[i] == separator:
                return i
        return -1
    
    def send_message(self, sock, data):
        """Send a message using the protocol: length + data"""
        length = len(data)

        if length < 0 or length > MAX_MESSAGE_SIZE:
            return None
        
        # Send length (4 bytes, big endian)
        length_bytes = self._int_to_bytes(length)
        sock.sendall(length_bytes) #sendall is guaranteed to send the entire message
        
        # Send data
        sock.sendall(data) #sendall is guaranteed to send the entire message
    
    def receive_message(self, sock):
        """Receive a message using the protocol: length + data"""
        
        # Receive length (4 bytes)
        length_data = self._recv_exact(sock, PAYLOAD_LENGTH_BYTES)
        if length_data is None:
            return None
        
        length = self._bytes_to_int(length_data)

        if length < 0 or length > MAX_MESSAGE_SIZE:
            return None
        
        # Receive message data
        message_data = self._recv_exact(sock, length)
        if message_data is None:
            return None
        
        return message_data
    
    def serialize_response(self, resp):
        """Convert Response to binary format"""
        # Response format: [success][separator][message]
        message_bytes = resp.message.encode('utf-8')
        total_size = RESPONSE_HEADER_SIZE + len(message_bytes)  # success + separator + message bytes
        
        buffer = bytearray(total_size)
        offset = 0
        
        # Write success flag
        buffer[offset] = 0x01 if resp.success else 0x00
        offset += 1
        
        # Write separator
        buffer[offset] = FIELD_SEPARATOR
        offset += 1
        
        # Write message
        buffer[offset:offset + len(message_bytes)] = message_bytes
        
        return bytes(buffer)

File: client/common/protocol.go
package common

import (
	"encoding/binary"
	"errors"
	"fmt"
	"io"
	"strconv"
)

const (
	// Protocol constants
	FIELD_SEPARATOR      = 0x00            // Null byte separator between fields
	BET_SEPARATOR        = 0xFF            // Separator between different bets
	MAX_MESSAGE_SIZE     = 1024 * 1024 * 2 // 2MB
	PAYLOAD_LENGTH_BYTES = 4               // 4 bytes for the length of the message
	RESPONSE_HEADER_SIZE = 2               // 2 bytes for the success flag and separator

	// Message types
	MESSAGE_TYPE_BET_BATCH         = 0x01
	MESSAGE_TYPE_FINISH_NOTIFY     = 0x02
	MESSAGE_TYPE_WINNERS_QUERY     = 0x03
	MESSAGE_TYPE_LOTTERY_COMPLETED = 0x04 // New message type for lottery completion notification

	// Response types (same as message types for now)
	RESPONSE_TYPE_BET_BATCH     = 0x01
	RESPONSE_TYPE_FINISH_NOTIFY = 0x02
	RESPONSE_TYPE_WINNERS_QUERY = 0x03
)

// Protocol handles binary communication protocol
type Protocol struct{}

// NewProtocol creates a new protocol instance
func NewProtocol() *Protocol {
	return &Protocol{}
}

// FinishNotification represents a notification that a client has finished sending all bets
type FinishNotification struct {
	AgencyID string
}

// WinnersQuery represents a query for winners of a specific agency
type WinnersQuery struct {
	AgencyID string
}

// WinnersResponse represents the response with winners for a specific agency
type WinnersResponse struct {
	Success bool
	Message string
	Winners []string // DNIs of winners
	Count   int
}

// LotteryCompletedNotification represents a notification that the lottery has been completed
type LotteryCompletedNotification struct {
	WinnersCount int
}

func writeAll(w io.Writer, b []byte) error {
	for len(b) > 0 {
		n, err := w.Write(b)
		if err != nil {
			return err
		}
		b = b[n:]
	}
	return nil
}

// SerializeBet converts BetData to binary format
func (p *Protocol) SerializeBet(bet *BetData) ([]byte, error) {
	log.Debugf("action: serialize_bet | result: in_progress | bet: %v", bet)

	// Calculate total size
	totalSize := p.calculateBetSize(bet)
	log.Debugf("action: serialize_bet | result: in_progress | totalSize: %v", totalSize)

	// Create buffer
	buffer := make([]byte, totalSize)
	offset := 0

	// Serialize the bet
	p.serializeBetFields(bet, buffer, &offset)

	log.Debugf("action: serialize_bet | result: success | buffer: %v", buffer)

	return buffer, nil
}

// SerializeBatch converts a slice of BetData to binary format
func (p *Protocol) SerializeBatch(bets []BetData) ([]byte, error) {
	log.Debugf("action: serialize_batch | result: in_progress | bets_count: %v", len(bets))

	// Calculate total size for all bets
	totalSize := 0
	for _, bet := range bets {
		totalSize += p.calculateBetSize(&bet) + 1 // +1 for bet separator
	}

	log.Debugf("action: serialize_batch | result: in_progress | totalSize: %v", totalSize)

	// Create buffer
	buffer := make([]byte, totalSize)
	offset := 0

	// Serialize each bet
	for _, bet := range bets {
		p.serializeBetFields(&bet, buffer, &offset)
		buffer[offset] = BET_SEPARATOR // Add bet separator
		offset++
	}

	log.Debugf("action: serialize_batch | result: success | buffer_size: %v", len(buffer))

	return buffer, nil
}

// calculateBetSize calculates the size needed to serialize a single bet
func (p *Protocol) calculateBetSize(bet *BetData) int {
	return len(bet.Nombre) + 1 + // nombre + separator
		len(bet.Apellido) + 1 + // apellido + separator
		len(bet.DNI) + 1 + // dni + separator
		len(bet.Nacimiento) + 1 + // nacimiento + separator
		len(bet.Numero) // numero (no final separator)
}

// serializeBetFields serializes the fields of a bet into the buffer at the given offset
func (p *Protocol) serializeBetFields(bet *BetData, buffer []byte, offset *int) {
	// Write nombre
	copy(buffer[*offset:], []byte(bet.Nombre))
	*offset += len(bet.Nombre)
	buffer[*offset] = FIELD_SEPARATOR
	*offset++

	// Write apellido
	copy(buffer[*offset:], []byte(bet.Apellido))
	*offset += len(bet.Apellido)
	buffer[*offset] = FIELD_SEPARATOR
	*offset++

	// Write dni
	copy(buffer[*offset:], []byte(bet.DNI))
	*offset += len(bet.DNI)
	buffer[*offset] = FIELD_SEPARATOR
	*offset++

	// Write nacimiento
	copy(buffer[*offset:], []byte(bet.Nacimiento))
	*offset += len(bet.Nacimiento)
	buffer[*offset] = FIELD_SEPARATOR
	*offset++

	// Write numero (no final separator)
	copy(buffer[*offset:], []byte(bet.Numero))
	*offset += len(bet.Numero)
}

// SendMessage sends a message using the protocol: length + data
func (p *Protocol) SendMessage(writer io.Writer, data []byte) error {
	length := uint32(len(data))
	log.Debugf("action: send_message | result: in_progress | length: %v", length)

	// Send length (4 bytes, big endian)
	lengthBytes := make([]byte, PAYLOAD_LENGTH_BYTES)
	binary.BigEndian.PutUint32(lengthBytes, length)
	log.Debugf("action: send_message | result: in_progress | lengthBytes: %v", lengthBytes)

	err := writeAll(writer, lengthBytes)
	if err != nil {
		return err
	}

	// Send data
	err = writeAll(writer, data)
	return err
}

// ReceiveMessage receives a message using the protocol: length + data
func (p *Protocol) ReceiveMessage(reader io.Reader) ([]byte, error) {
	// Receive length (4 bytes)
	lengthBytes := make([]byte, PAYLOAD_LENGTH_BYTES)
	log.Debugf("action: receive_message | result: in_progress | lengthBytes: %v", lengthBytes)
	_, err := io.ReadFull(reader, lengthBytes) //readfull is guaranteed to read the entire message
	if err != nil {
		return nil, err
	}

	length := binary.BigEndian.Uint32(lengthBytes)
	log.Debugf("action: receive_message | result: in_progress | length: %v", length)

	// Receive message data
	messageBytes := make([]byte, length)
	_, err = io.ReadFull(reader, messageBytes) //readfull is guaranteed to read the entire message
	if err != nil {
		return nil, err
	}

	return messageBytes, nil
}

// Response represents server response
type Response struct {
	Success bool
	Message string
}

// DeserializeResponse converts binary data to Response
func (p *Protocol) DeserializeResponse(data []byte) (*Response, error) {
	if len(data) < RESPONSE_HEADER_SIZE {
		return nil, errors.New("response data too short")
	}

	resp := &Response{}

	// Read success flag
	resp.Success = data[0] == 0x01 // 0x01 for success, 0x00 for failure

	// Skip separator
	if data[1] != FIELD_SEPARATOR {
		return nil, errors.New("invalid response format")
	}

	// Read message (rest of the data)
	resp.Message = string(data[2:])

	return resp, nil
}

// SerializeFinishNotification converts FinishNotification to binary format
func (p *Protocol) SerializeFinishNotification(notification *FinishNotification) ([]byte, error) {
	// Calculate total size: agency_id + separator
	totalSize := len(notification.AgencyID) + 1

	buffer := make([]byte, totalSize)
	offset := 0

	// Write agency ID
	copy(buffer[offset:], []byte(notification.AgencyID))
	offset += len(notification.AgencyID)

	// Write separator
	buffer[offset] = FIELD_SEPARATOR

	return buffer, nil
}

// SerializeWinnersQuery converts WinnersQuery to binary format
func (p *Protocol) SerializeWinnersQuery(query *WinnersQuery) ([]byte, error) {
	// Calculate total size: agency_id + separator
	totalSize := len(query.AgencyID) + 1

	buffer := make([]byte, totalSize)
	offset := 0

	// Write agency ID
	copy(buffer[offset:], []byte(query.AgencyID))
	offset += len(query.AgencyID)

	// Write separator
	buffer[offset] = FIELD_SEPARATOR

	return buffer, nil
}

// DeserializeWinnersResponse converts binary data to WinnersResponse
func (p *Protocol) DeserializeWinnersResponse(data []byte) (*WinnersResponse, error) {
	if len(data) < RESPONSE_HEADER_SIZE {
		return nil, errors.New("response data too short")
	}

	resp := &WinnersResponse{}

	// Read success flag
	resp.Success = data[0] == 0x01 // 0x01 for success, 0x00 for failure

	// Skip separator
	if data[1] != FIELD_SEPARATOR {
		return nil, errors.New("invalid response format")
	}

	// Parse the rest of the data
	remainingData := string(data[2:])

	// Split by field separator to get count and winners
	parts := []byte(remainingData)

	// Find the count field (first field after success)
	offset := 0
	countStr := ""
	for offset < len(parts) && parts[offset] != FIELD_SEPARATOR {
		countStr += string(parts[offset])
		offset++
	}

	if offset < len(parts) {
		offset++ // Skip separator
	}

	// Parse winners (DNIs separated by FIELD_SEPARATOR)
	winners := []string{}
	currentWinner := ""
	for offset < len(parts) {
		if parts[offset] == FIELD_SEPARATOR {
			if currentWinner != "" {
				winners = append(winners, currentWinner)
				currentWinner = ""
			}
		} else {
			currentWinner += string(parts[offset])
		}
		offset++
	}

	// Add last winner if exists
	if currentWinner != "" {
		winners = append(winners, currentWinner)
	}

	resp.Winners = winners
	resp.Count = len(winners)
	resp.Message = fmt.Sprintf("Found %d winners", resp.Count)

	return resp, nil
}

// DeserializeLotteryCompletedNotification converts binary data to LotteryCompletedNotification
func (p *Protocol) DeserializeLotteryCompletedNotification(data []byte) (*LotteryCompletedNotification, error) {
	if len(data) < 1 {
		return nil, errors.New("notification data too short")
	}

	// Find the separator
	separatorPos := -1
	for i, b := range data {
		if b == FIELD_SEPARATOR {
			separatorPos = i
			break
		}
	}

	if separatorPos == -1 {
		return nil, errors.New("invalid lottery completed notification format")
	}

	// Extract winners count
	countStr := string(data[:separatorPos])
	count, err := strconv.Atoi(countStr)
	if err != nil {
		return nil, fmt.Errorf("invalid winners count: %v", err)
	}

	return &LotteryCompletedNotification{WinnersCount: count}, nil
}

// SendMessageWithType sends a message with type header using the protocol: type + length + data
func (p *Protocol) SendMessageWithType(writer io.Writer, messageType byte, data []byte) error {
	length := uint32(len(data))
	log.Debugf("action: send_message_with_type | result: in_progress | type: %v | length: %v", messageType, length)

	// Send message type (1 byte)
	typeBytes := []byte{messageType}
	err := writeAll(writer, typeBytes)
	if err != nil {
		return err
	}

	// Send length (4 bytes, big endian)
	lengthBytes := make([]byte, PAYLOAD_LENGTH_BYTES)
	binary.BigEndian.PutUint32(lengthBytes, length)
	log.Debugf("action: send_message_with_type | result: in_progress | lengthBytes: %v", lengthBytes)

	err = writeAll(writer, lengthBytes)
	if err != nil {
		return err
	}

	// Send data
	err = writeAll(writer, data)
	return err
}

// ReceiveMessageWithType receives a message with type header using the protocol: type + length + data
func (p *Protocol) ReceiveMessageWithType(reader io.Reader) (byte, []byte, error) {
	// Receive message type (1 byte)
	typeBytes := make([]byte, 1)
	_, err := io.ReadFull(reader, typeBytes)
	if err != nil {
		return 0, nil, err
	}

	messageType := typeBytes[0]
	log.Debugf("action: receive_message_with_type | result: in_progress | type: %v", messageType)

	// Receive length (4 bytes)
	lengthBytes := make([]byte, PAYLOAD_LENGTH_BYTES)
	_, err = io.ReadFull(reader, lengthBytes)
	if err != nil {
		return 0, nil, err
	}

	length := binary.BigEndian.Uint32(lengthBytes)

	// Receive message data
	messageBytes := make([]byte, length)
	_, err = io.ReadFull(reader, messageBytes)
	if err != nil {
		return 0, nil, err
	}

	return messageType, messageBytes, nil
}
